### PR TITLE
address: add AddressTaproot

### PR DESCRIPTION
**Pull request by guggero**
Fixes https://github.com/btcsuite/btcutil/issues/200.

### CODE REVIEW

**Comment by guggero** [comment]
Rebased and polished my old code I had lying around for this. PTAL @Roasbeef

File: address_test.go [hunk 1]
```diff
@@ -655,13 +655,36 @@ func TestAddresses(t *testing.T) {
...
 		},
-		// Unsupported witness versions (version 0 only supported at this point)
+
+		// P2TR address tests.
 		{
-			name:  "segwit mainnet witness v1",
-			addr:  "bc1pw508d6qejxtdg4y5r3zarvary0c5xw7kw508d6qejxtdg4y5r3zarvary0c5xw7k7grplx",
-			valid: false,
-			net:   &chaincfg.MainNetParams,
+			name:    "segwit v1 mainnet p2tr",
```
Comment (Roasbeef) [inline address_test.go:661]: Let's also add these test vectors? https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki#test-vectors-for-v0-v16-native-segregated-witness-addresses

File: address_test.go [hunk 2]
```diff
@@ -655,13 +655,36 @@ func TestAddresses(t *testing.T) {
 			},
 			net: &customParams,
 		},
-		// Unsupported witness versions (version 0 only supported at this point)
+
+		// P2TR address tests.
```
Comment (Roasbeef) [inline address_test.go:659]: I think we also want test vectors for the negative reject case: someone tries to use a segwit v1 witness program w/ the OG bech32 version/checksum

**Comment by Roasbeef** [comment]
Once this is in, we'll also want to update this method as well: https://pkg.go.dev/github.com/btcsuite/btcd/txscript#PayToAddrScript 

As well as this type: https://pkg.go.dev/github.com/btcsuite/btcd/txscript#PkScript

File: address.go [hunk 3]
```diff
@@ -149,8 +161,9 @@ func DecodeAddress(addr string, defaultNet *chaincfg.Params) (Address, error) {
 			}
 
 			// We currently only support P2WPKH and P2WSH, which is
-			// witness version 0.
-			if witnessVer != 0 {
+			// witness version 0 and P2TR which is witness version
```
Comment (Roasbeef) [inline address.go:164]: We want to also have the `decodeSegWitAddress` method return the `bech32` version returned as well. This way we can reject decoding an address that's segwit v1, but that doesn't use the `bech32m` encoding.

File: address.go [hunk 4]
```diff
@@ -681,3 +698,87 @@ func (a *AddressWitnessScriptHash) WitnessVersion() byte {
 func (a *AddressWitnessScriptHash) WitnessProgram() []byte {
 	return a.witnessProgram[:]
 }
+
+// AddressTaproot is an Address for a pay-to-taproot (P2TR) output. See BIP 341
+// for further details.
+type AddressTaproot struct {
+	hrp            string
```
Comment (Roasbeef) [inline address.go:705]: Extract this into a base `AddressSegwit` struct so we can save on the duplicaation here and above? This way it can also inherit all the shared methods used here.

File: address.go (hunk 3 above)
Comment (Roasbeef) [reply address.go:164]: To accomplish this, we need to change the method to instead use `DecodeGeneric` so it can obtain the `bech32` version used when encoding.

**Review by Roasbeef — CHANGES_REQUESTED** [review]

File: address_test.go (hunk 1 above)
Comment (guggero) [reply address_test.go:661]: Yeah, added all of them. Also fixed some edge cases covered by those test vectors.

File: address_test.go (hunk 2 above)
Comment (guggero) [reply address_test.go:659]: Right, this was made apparent by the test cases as well :+1:

File: address.go (hunk 4 above)
Comment (guggero) [reply address.go]: Makes sense, was able to remove a lot of duplicated code!

**Review by ellemouton — APPROVED** [review]
LGTM 💥  Much excite!
